operators/pkg/controller/elasticsearch/driver/upgrade_test.go

```diff
@@ -0,0 +1,315 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package driver
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/reconciler"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/nodespec"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/sset"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/utils/stringsutil"
+	"github.com/go-test/deep"
+	"github.com/stretchr/testify/require"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+type mockESState struct {
+	shardAllocationsEnabled bool
+	green                   bool
+	nodeNames               []string
+}
+
+func (m mockESState) NodesInCluster(nodeNames []string) (bool, error) {
+	return stringsutil.StringsInSlice(nodeNames, m.nodeNames), nil
+}
+
+func (m mockESState) ShardAllocationsEnabled() (bool, error) {
+	return m.shardAllocationsEnabled, nil
+}
+
+func (m mockESState) GreenHealth() (bool, error) {
+	return m.green, nil
+}
+
+var _ ESState = mockESState{}
+
+var defaultESState = mockESState{
+	shardAllocationsEnabled: false,
+	green:                   true,
+	nodeNames:               []string{"default-0"},
+}
+
+type mockUpdater map[string]int32
+
+func (m mockUpdater) updatePartition(sset *appsv1.StatefulSet, newPartition int32) error {
+	m[sset.Name] = newPartition
+	return nil
+}
+
+func newResults() *reconciler.Results {
+	return &reconciler.Results{}
+}
+
+func Test_defaultDriver_doRollingUpgrade(t *testing.T) {
+	// This does not test: podUpgradeDone or prepareClusterForNodeRestart in detail but is focused on the main invariants
```

Comment:
Is this something to be addressed in an other PR ?
I'm 👍 with the tests, my only concern is that there is no test with a mix of "upgraded" and "yet to be upgraded" pods.

Reply:
+1 on adding tests that cover some specifics of podUpgradeDone. Totally fine in another PR (and can be done by someone else).

Reply:
No worries happy to spend a little more time on it